File: hash/leetcode_49/solution.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const anagramMap = new Map();

    for (const str of strs) {
        const key = str.split('').sort().join('');

        if (anagramMap.has(key)) {
            anagramMap.get(key).push(str);
        } else {
            anagramMap.set(key, [str]);
        }
    } 

    return [...anagramMap.values()];
};

File: hash/leetcode_49/solution.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_groups = defaultdict(list)

        for str in strs:
            key = "".join(sorted(str))

            anagram_groups[key].append(str)

        return list(anagram_groups.values())